util/conversion/conversion.go

```diff
@@ -54,36 +56,34 @@ var (
 // the Custom Resource Definition and looks which one is the stored version available.
 //
 // The object passed as input is modified in place if an updated compatible version is found.
-func ConvertReferenceAPIContract(ctx context.Context, c client.Client, ref *corev1.ObjectReference) error {
+func ConvertReferenceAPIContract(ctx context.Context, log logr.Logger, c client.Client, restConfig *rest.Config, ref *corev1.ObjectReference) error {
 	gvk := ref.GroupVersionKind()
-	crd, err := util.GetCRDWithContract(ctx, c, gvk, contract)
+
+	metadata, err := util.GetCRDMetadataFromGVK(ctx, restConfig, gvk)
 	if err != nil {
-		return err
+		log.Info("Cannot retrieve CRD with metadata only client, falling back to slower listing", "err", err.Error())
+		// Fallback to slower and more memory intensive method to get the full CRD.
+		crd, err := util.GetCRDWithContract(ctx, c, gvk, contract)
+		if err != nil {
+			return err
+		}
+		metadata = &metav1.PartialObjectMetadata{
+			TypeMeta:   crd.TypeMeta,
+			ObjectMeta: crd.ObjectMeta,
+		}
 	}
 
 	// If there is no label, return early without changing the reference.
-	supportedVersions, ok := crd.Labels[contract]
+	supportedVersions, ok := metadata.Labels[contract]
 	if !ok || supportedVersions == "" {
-		return errors.Errorf("cannot find any versions matching contract %q for CRD %v", contract, crd.Name)
+		return errors.Errorf("cannot find any versions matching contract %q for CRD %v", contract, metadata.Name)
 	}
 
 	// Pick the latest version in the slice and validate it.
 	kubeVersions := util.KubeAwareAPIVersions(strings.Split(supportedVersions, "_"))
 	sort.Sort(kubeVersions)
 	chosen := kubeVersions[len(kubeVersions)-1]
 
-	// Validate that the picked version is actually in the CRD spec.
```

Comment:
Working with metadata we are losing the validation that the picked version is an actual version, but I think this is an acceptable trade-off